R/expect_loglinear_time.R

```diff
@@ -0,0 +1,18 @@
+#' Function to test if input algorithm has log-linear time complexity
+#'
+#' @title Log-Linear Time Complexity Test function
+#'
+#' @param f A function which returns a data frame composed of timings and corresponding data sizes, ideally from asymptoticTimings. (can use other functions)
+#'
+#' @param ... Parameters for passed function 'f'.
+#'
+#' @return null, throws error if complexity is not log-linear. (i.e. O(NlogN) for size N, in Big-O notation)
+#'
+#' @export
+
+expect_loglinear_time = function(f = asymptoticTimings, ...)
+{
+  timings.df <- f(...)
+
+  stopifnot(asymptoticTimeComplexityClass(timings.df) == "loglinear")
```

Comment:
since the only difference between this function and the other expect_*_time functions is the complexity class, you should make a generic funciton like 
```
expect_time_complexity <- function(complexity.class, f){......}
```
and then use it from the other expect functions like
```
expect_loglinear_time <- function(...)expect_time_complexity("loglinear", ...)
```
  expect_time_complexity(..., c

Reply:
If I get you right then it would mean `expect_time_complexity` would contain the complexity comparison part (to avoid that duplication) and it would be internally called in each of the different complexity classes' expect functions with the corresponding class, right?  

But then wouldn't a single function with the added `complexity.class` parameter work?
```r
expect_time_complexity = function(complexity.class, f, ...)
{
  timings.df <- f(...)
  if(asymptoticTimeComplexityClass(timings.df) != complexity.class)
    stop("Complexity mismatch: Expected ", complexity.class, " complexity")
}
```
Also would it be better to return the string (can use `invisible()` in conjunction as well) denoting the complexity class in the complexity classifying functions instead of printing it? It doesn't seem to provide any issue, but it stacks up, as visible in a RCMD `check()`.  <br>
Can also use:
```r
expect_time_complexity = function(expected.complexity.class, f, ...)
{
  timings.df <- f(...)
  predicted.complexity.class <- asymptoticTimeComplexityClass(timings.df)
  if(predicted.complexity.class != expected.complexity.class)
    stop("Complexity mismatch: Expected ", expected.complexity.class, " complexity, in place of the predicted ", predicted.complexity.class, " complexity")
}
```

Reply:
actually instead of stop you should use expect_identical or some other expect_ function (maybe write your own?) which gives an informative error message.

Reply:
btw what should the error message contain to be informative? there aren't any tracebacks when I use `expect_xyz` functions, so am wondering is it just a casual message regarding string mismatch, or something which a `try()` produces?